Add missing DECL and DEAD instructions in function pointer call site labelling

remi-delmas-3000 opened the pull request:
No change in functionality or performance is expected. 

Previously, function pointer call site labelling introduced ASSIGNS and CALL instructions to a fresh function pointer variable in the goto program without adding corresponding DECL and DEAD instructions, which made them look like global variables when they really are local to the function.

This caused  problems with function contracts checking, because a function would appear to be writing to a global that is not mentioned in its assigns clause.

We also add a small diagram to document what it means for `goto_programt::insert_before_swap` to preserve jumps to target.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

tautschnig commented on src/goto-programs/goto_program.h:
````diff
@@ -625,6 +625,22 @@ class goto_programt
 
   /// Insertion that preserves jumps to "target".
   /// The instruction is destroyed.
+  ///
+  /// Turns:
+  /// ```
+  /// ...->[a]->...
+  ///       ^
+  ///     target
+  /// ```
+  ///
+  /// Into:
+  /// ```
+  /// ...->[i]->[a]->...
+  ///       ^
+  ///     target
+  /// ```
+  ///
+  /// So that jumps to `a` now jump to the newly inserted `i`.
````
Thank you! But then: would you mind putting this in a commit of its own?

tautschnig commented on src/goto-programs/label_function_pointer_call_sites.cpp:
```diff
@@ -52,17 +52,31 @@ void label_function_pointer_call_sites(goto_modelt &goto_model)
+        auto decl_instruction =
+          goto_programt::make_decl(new_function_pointer, source_location);
+        goto_function.second.body.insert_before_swap(it, decl_instruction);
+        it++;
```
Nit picking: `++it` (which the compiler will likely take care of for you, but why rely on compiler optimisations...).

tautschnig commented on src/goto-programs/label_function_pointer_call_sites.cpp:
```diff
@@ -52,17 +52,31 @@ void label_function_pointer_call_sites(goto_modelt &goto_model)
+        it++;
+
+        // transform original call into a call to the new variable
+        to_code_function_call(it->code_nonconst()).function() =
```
`it->call_function() =` would be a shorter version thereof.

tautschnig reviewed (approved):
Since you touched this code, would you mind adding the following cleanup:
1) The lambda inserting the symbol into the symbol table should also set `is_thread_local` and `is_file_local` to `true`.
2) Instead of the repeated and long-winded `goto_model.symbol_table.lookup_ref` could you please introduce a `namespacet ns{goto_model.symbol_table}` early on and replace the `goto_model.symbol_table.lookup_ref` by `ns.lookup`?

Also, would it be possible to include a test, for example by adding patterns to `regression/goto-instrument/restrict-function-pointer-by-name-local/test.desc` that check for the existence of the new DECL/DEAD instructions?

Commit:
Add DECL and DEAD instructions when labelling function pointer call sites.

Previously, function pointer call site labelling introduced ASSIGNS and CALL instructions
to a fresh function pointer variable in the goto program without adding corresponding
DECL and DEAD instructions which made them look like global variables when they really
are local to the function.

Commit:
Adding a diagram to the doc of `goto_programt::insert_before_swap` to illustrate "preserves jumps to target".
